### PR TITLE
feat(sources/file): fall back to global data_dir option (#644)

### LINKED ISSUE #644: Require the `data_dir` option for the `file` source and default it to the global `data_dir`
Currently, the `file` source requires a `data_dir` for checkpointing, yet this requirement is not enforced when Vector is started. You can see [here](https://github.com/timberio/vector/blob/b420159287f19a2aa4405da6f90fcea733d9de28/src/sources/file.rs#L44) that the `data_dir` option defaults to an empty path. This should not be allowed since an explicit data directory should be specified to store the checkpoint positions.

### Requirements

- [ ] Default the `file` source `data_dir` option to the global `data_dir`
- [ ] Exit on start if the `file` source `data_dir` option is empty.
- [ ] Allow the user to pass a `data_dir` option in the file source itself to override the global `data_dir`.

### PR DESCRIPTION
**Pull request by seeekr**
### Description

Closes #644.

There is not a `GlobalOptions` struct that should contain any, well, globally relevant options that any sources, sinks or transforms might be interested in when launching their tasks from the configuration they are given.

So far only the `SourceConfig` trait has been adjusted to accept the new, additional argument in the `build` method.

`data_dir`, set via the config file's global section, is the single newly available global option.

I've considered also adding a `config_path: PathBuf` field, as something like that is often useful in a configuration context (i.e. support for discovering additional file resources in the future relative to the global config file, think "config.d" folder etc). But it seemed that would require changing signatures of several methods, so it wasn't a no-brainer to do that.

I did not include any tests yet. A little uncertain as to whether it's worth it as it's basically a trivial bit of code, and it's not as straightforward to test it as the code's effect is not directly observable with simple code. At least I didn't immediately think of a way to do it easily. (I did not think very hard yet, I have to say.) Eager to hear how y'all think about testing in this project, with this change as a (very small) example.

I'm not a fan of how much code I had to change in order to make this small change. Maybe changing the `SourceConfig::build` method was OK, but then there's still a lot of repetition in the test code. Ideally we'd clean that up by providing some macros and/or more test helper methods. It looks repetitive, though I've looked only very briefly into whether I could spot a pattern in that code (especially in sources/file.rs) that would allow for an easy fix via macro.

I think that's all. I've kept all changes relevant to #644 in a single commit, so squashing is hopefully not necessary. That'd allow the other commits to survive, which I think would be preferable.

I did not update any documentation. Please advise (@binarylogic) if I should update anything and where. I'm assuming a small update of any section talking about configuring data_dirs for file sources might be necessary.

### TODO

- [x] All commits are signedoff (See CONTRIBUTING.md). Forgot? `make signoff`.
- [x] CHANGELOG.md has been updated to reflect noteworthy changes.
- [x] Documentation has been updated to reflect changes (see DOCUMENTING.md).

### CODE REVIEW

**Comment by seeekr** [comment]
And tests seem to pass on my machine, except for aws/kinesis ones. I'm assuming that's not related.

File: src/sources/file.rs [hunk 1]
```diff
@@ -44,25 +44,51 @@ impl Default for FileConfig {
...
+    ) -> Result<super::Source, String> {
+        let data_dir = match &self.data_dir {
+            Some(v) => v.clone(),
+            None => match &globals.data_dir {
+                Some(v) => v.clone(),
+                None => {
+                    return Err("data_dir option required, but not given here or globally".into())
+                }
+            },
+        };
```
[the remark this answers is not in the document]
Comment (seeekr) [reply src/sources/file.rs]: You're right, will improve that.

[the remark this answers is not in the document]
Comment (seeekr) [reply src/sources/file.rs]: Should be fixed now.

**Comment by seeekr** [comment]
Thanks for the review comments. There was indeed a simple way to test the new code -- extracting into its own method and testing that :)

**Comment by binarylogic** [comment]
@seeekr nice work on this. Once you get an approval feel free to merge whenever you're ready.

**Comment by lukesteensen** [comment]
Looks good! If you'd like to maintain your independent commits, feel free to do the merge from the command line.

**Comment by bittrance** [comment]
I think this PR will introduce a bug in its current shape. PR #609 checkpointing feature assumes that it is safe to delete the `checkpoints` dir. However, if there are several `file` source entries in the configuration with the same `data_dir`, there will be several checkpointers running, all wanting to delete the whole `checkpoints` dir, meaning they will throw away all other checkpoints and write their own 🐛 . You can observe this issue on master today if you create two `file` sources with identical `data_dir`. Stracing:
```
[pid 15886] rmdir("/var/lib/vector/checkpoints") = 0
[pid 15886] mkdir("/var/lib/vector/checkpoints", 0777) = 0
[pid 15886] rmdir("/var/lib/vector/checkpoints") = 0
[pid 15886] mkdir("/var/lib/vector/checkpoints", 0777) = 0
[pid 15884] mkdir("/var/lib/vector/checkpoints", 0777 <unfinished ...>
[pid 15885] rmdir("/var/lib/vector/checkpoints" <unfinished ...>
[pid 15884] <... mkdir resumed> )       = -1 EEXIST (File exists)
[pid 15885] <... rmdir resumed> )       = 0
[pid 15885] mkdir("/var/lib/vector/checkpoints", 0777
Jul 24 10:34:33.810 WARN source{name="source__burt_staging-eu1_hub_africa_staging_log"}:file-server: file_source::file_server: Problem writing checkpoints: Os { code: 17, kind: AlreadyExists, message: "File exists" }
) = 0
[pid 15886] rmdir("/var/lib/vector/checkpoints") = 0
[pid 15886] mkdir("/var/lib/vector/checkpoints", 0777) = 0
[pid 15885] mkdir("/var/lib/vector/checkpoints", 0777) = -1 EEXIST (File exists)
[pid 15884] rmdir("/var/lib/vector/checkpoints") = 0
[pid 15884] mkdir("/var/lib/vector/checkpoints", 0777) = 0
[pid 15886] rmdir("/var/lib/vector/checkpoints") = 0
[pid 15886] mkdir("/var/lib/vector/checkpoints", 0777) = 0
[pid 15886] rmdir("/var/lib/vector/checkpoints") = 0
[pid 15886] mkdir("/var/lib/vector/checkpoints", 0777) = 0
[pid 15885] rmdir("/var/lib/vector/checkpoints" <unfinished ...>
[pid 15884] rmdir("/var/lib/vector/checkpoints") = 0
[pid 15884] mkdir("/var/lib/vector/checkpoints", 0777) = 0
[pid 15885] <... rmdir resumed> )       = 0
[pid 15885] mkdir("/var/lib/vector/checkpoints", 0777
Jul 24 10:34:37.910 WARN source{name="source__burt_staging-eu1_hub_africa_staging_log"}:file-server: file_source::file_server: Problem writing checkpoints: Os { code: 2, kind: NotFound, message: "No such file or directory" }
) = 0
```

**Comment by lukesteensen** [comment]
Good catch, thank you @bittrance! I think the proper solution is probably to add a layer of namespacing to the existing `data_dir` based on the name of the source. So we can keep the existing logic, but then basically do `data_dir.join(format!("{}_checkpoints", source_name))`.

I think the biggest complication here is that sources don't get passed their own name, so we'll need to come up with a solution for that too. What do you think @seeekr?

**Comment by bittrance** [comment]
@lukesteensen As far as I understand when you set a source-local `data_dir` (or namespace as you propose), vector will treat `/var/log/messages` as a new file when I rename its including source declaration or move its `include` between source declarations, e.g. from `[source.all_the_files]` to `[source.application_logs]` and `[source.system_logs]`. This may lead to log line duplication. I don't think the user expects that. I think a better solution would be to have a single checkpointer shared by all `file` sources. That way, memory is retained for a single file as long as the checkpointing info is on disk.

**Comment by lukesteensen** [comment]
@bittrance My concern with a shared checkpointer would be conflicts in the case that multiple sources include the same file. The way vector is designed, sources of the same type are still intended to be totally independent of each other.

There might be a way that we could have file sources with no checkpoint saved for a given file look for checkpoints from other file sources using the same data dir and use that as a starting point, but I'm not sure it'd be worthwhile. If users are configuring multiple file sources, there may be the expectation that they operate independently. Otherwise, I think our general recommendation would be to use a single file source watching multiple paths.

Any thoughts @binarylogic?

**Comment by binarylogic** [comment]
Yes, thank you @bittrance. We will definitely solve that case. And I agree with @lukesteensen's [last comment](https://github.com/timberio/vector/pull/673#issuecomment-514688832) that sources should operate independently. Especially since I could see users setting up multiple disparate pipelines.

**Comment by seeekr** [comment]
@lukesteensen I think your suggested change will keep the system sound, without introducing much change. I'd just not mess with the "checkpoints" dir name, since that is set in the Checkpointer inside the file-source crate, and instead set the data_dir to `"{data_dir}/source_{source_name}"`, i.e. subdir instead of prefix/suffix on the checkpoints dir. Feels a little cleaner to me. WDYT?

**Comment by seeekr** [comment]
And basically we'd need to add one more argument (the source name) to the `build` method. (The only regret is that Rust is not at Java-like levels of refactoring support, so this change takes minutes instead of seconds to perform.)

**Comment by seeekr** [comment]
Iiuc that it's only the checkpointer that needs the data_dir here, then the remaining issue should now be taken care of. PTAL, @lukesteensen!

**Pull request merged**